src/printer.js

```diff
@@ -3248,6 +3283,118 @@ function hasLeadingOwnLineComment(text, node) {
   return res;
 }
 
+function hasNakedLeftSide(node) {
+  return (
+    node.type === "AssignmentExpression" ||
+    node.type === "BinaryExpression" ||
+    node.type === "LogicalExpression" ||
+    node.type === "ConditionalExpression" ||
+    node.type === "CallExpression" ||
+    node.type === "MemberExpression" ||
+    node.type === "SequenceExpression" ||
+    node.type === "TaggedTemplateExpression"
+  );
+}
+
+function getLeftSide(node) {
+  if (node.expressions) {
+    return node.expressions[0];
+  }
+  return node.left || node.test || node.callee || node.object || node.tag;
+}
+
+function exprNeedsASIProtection(node) {
+  // HACK: node.needsParens is added in `genericPrint()` for the sole purpose
+  // of being used here. It'd be preferable to find a cleaner way to do this.
+  const maybeASIProblem = node.needsParens ||
```

Comment:
I see that this is recursive, and the top-level statement kicks this off. I'm assuming it's possible for a node to be mutated deep down when printing, and then the recursion here eventually picks that up? I thought at first you only need to check top-level statements, but I think it's more complex than that.

As I said above I'm not too worried about merging this as is, and if we ever get around to cleaning it up, we can. We should keep this in mind if we ever encourage people to pass ASTs straight in though, as we will be modifying it (though we modify it with comments anyway).